![](../storage/banners/40_assumptions.png)

`import kiamformation as kf`

In [1]:
from __init__ import *

Инициализация проекта kiam-formation | Контекст: kiamformation


##### Быстродействие

`clip`: своё лучше для обычных чисел

In [43]:
import numpy as np
import time

n = int(1e6)

# Моя функция
def clip(a: float, bot: float, top: float) -> float:
    if a < bot:
        return bot
    if a > top:
        return top
    return a

t0 = time.time()
for _ in range(n):
    tmp = clip(-1, 0, 2)
print(f"my-clip: {time.time() - t0} секунд")

t0 = time.time()
for _ in range(n):
    tmp = np.clip(-1, 0, 2)
print(f"np.clip: {time.time() - t0} секунд")

my-clip: 0.14356112480163574 секунд
np.clip: 3.1530544757843018 секунд


##### Кватернионы в ПО

In [ ]:
# import numpy as np
# s0 = dir(np)

In [ ]:
# s = dir(quaternion.np)
# for i in s0:
#     s.remove(i)
# s

**Инициализация**

In [18]:
import numpy as np
import quaternion

for q in [np.quaternion(1, 0, 0, 0),
          np.quaternion(1, 2, 3, 4),
          np.quaternion(5, 0, 0),
          np.quaternion(5),
          np.quaternion()]:
    print(f"q = {q}")

q = np.quaternion(1, 2, 3, 4)

q = quaternion(1, 0, 0, 0)
q = quaternion(1, 2, 3, 4)
q = quaternion(0, 5, 0, 0)
q = quaternion(5, 0, 0, 0)
q = quaternion(0, 0, 0, 0)


In [28]:
a = np.array([0, 0, 0.5])

q1 = quaternion.from_vector_part(a)
q2 = quaternion.from_euler_angles(a)

q3 = np.quaternion(np.sqrt(1 - np.linalg.norm(a)**2), *a)

print(f"Кватернион из вектора: {q1}")
print(f"Кватернион единичный из вектора: {q3}")
print(f"Кватернион из углов Эйлера: {q2}")

Кватернион из вектора: quaternion(0, 0, 0, 0.5)
Кватернион единичный из вектора: quaternion(0.866025403784439, 0, 0, 0.5)
Кватернион из углов Эйлера: quaternion(0.968912421710645, 0, 0, 0.247403959254523)


**Преобразования типов**

In [66]:
a = quaternion.as_float_array(q)
b = quaternion.np.array(q)
c = quaternion.np.asarray(q)
d = q.components

print(f"Как numpy-массив: {a}")
print(type(a))
print(f"Как numpy-массив: {b}")
print(type(b))
print(f"Как numpy-массив: {c}")
print(type(c))
print(f"Как numpy-массив: {d}")
print(type(d))

Как numpy-массив: [1. 2. 3. 4.]
<class 'numpy.ndarray'>
Как numpy-массив: quaternion(1, 2, 3, 4)
<class 'numpy.ndarray'>
Как numpy-массив: quaternion(1, 2, 3, 4)
<class 'numpy.ndarray'>
Как numpy-массив: [1. 2. 3. 4.]
<class 'numpy.ndarray'>


In [132]:
a = np.array([4, 5, 6, 7])

anw1 = np.append(a, q)
anw2 = np.append(a, q.components)

print(f"Плохой пример: {anw1}")
print(f"Хороший пример: {anw2}")

Плохой пример: [quaternion(4, 0, 0, 0) quaternion(5, 0, 0, 0) quaternion(6, 0, 0, 0)
 quaternion(7, 0, 0, 0) quaternion(1, 2, 3, 4)]
Хороший пример: [4. 5. 6. 7. 1. 2. 3. 4.]


**Преобразовани математические**

In [101]:
a = q.real
b = q.vec
w = q.w
x = q.x
y = q.y
z = q.z

print(f"Скалярная часть: {a}")
print(type(a))
print(f"Векторная часть: {b}")
print(type(b))
print(f"Составлющие: {w} {x} {y} {z}")
print(type(w), type(x), type(y), type(z))

Скалярная часть: 1.0
<class 'float'>
Векторная часть: [2. 3. 4.]
<class 'numpy.ndarray'>
Составлющие: 1.0 2.0 3.0 4.0
<class 'float'> <class 'float'> <class 'float'> <class 'float'>


In [103]:
a = quaternion.np.normalized(q)
b = q.normalized()

print(f"Нормированный кватернион: {a} | {b}")
print(type(a), type(b))

Нормированный кватернион: quaternion(0.182574185835055, 0.365148371670111, 0.547722557505166, 0.730296743340221) | quaternion(0.182574185835055, 0.365148371670111, 0.547722557505166, 0.730296743340221)
<class 'quaternion.quaternion'> <class 'quaternion.quaternion'>


In [99]:
a = quaternion.np.abs(q)
b = quaternion.np.absolute(q)
d = q.abs()

c = quaternion.np.norm(q)

print(f"Модуль: {a} | {b} | {d}")
print(type(a), type(b), type(d), '\n')

print(f"Норма: {c} (√c = {np.sqrt(c)})")
print(type(c))

Модуль: 5.477225575051661 | 5.477225575051661 | 5.477225575051661
<class 'numpy.float64'> <class 'numpy.float64'> <class 'float'> 

Норма: 30.0 (√c = 5.477225575051661)
<class 'numpy.float64'>


In [27]:
a = quaternion.np.angle_of_rotor(q)
b = q.angle()

print(f"Угол поворота: {a} | {b}")
print(type(a), type(b))

Угол поворота: 4.389231563842282 | 4.389231563842282
<class 'numpy.float64'> <class 'float'>


In [96]:
a = q.conj()
b = q.conjugate()

с = q.inverse()
d = q.reciprocal()

print(f"Сопряженный: {a} | {b}")
print(type(a), type(b), '\n')

print(f"Обратный: {с} | {d}")
print(type(с), type(d))

Сопряженный: quaternion(1, -2, -3, -4) | quaternion(1, -2, -3, -4)
<class 'quaternion.quaternion'> <class 'quaternion.quaternion'> 

Обратный: quaternion(0.0333333333333333, -0.0666666666666667, -0.1, -0.133333333333333) | quaternion(0.0333333333333333, -0.0666666666666667, -0.1, -0.133333333333333)
<class 'quaternion.quaternion'> <class 'quaternion.quaternion'>


In [125]:
A = quaternion.as_rotation_matrix(q)
B = quaternion.as_rotation_matrix(q.normalized())

print(f"Матрица поворота из кватерниона A: \n{A}")
print(f"Матрица поворота из кватерниона B: \n{B}")

Матрица поворота из кватерниона A: 
[[-0.66666667  0.13333333  0.73333333]
 [ 0.66666667 -0.33333333  0.66666667]
 [ 0.33333333  0.93333333  0.13333333]]
Матрица поворота из кватерниона B: 
[[-0.66666667  0.13333333  0.73333333]
 [ 0.66666667 -0.33333333  0.66666667]
 [ 0.33333333  0.93333333  0.13333333]]


##### Проверка кватернионного умножения

###### Символьные выражения

In [3]:
from sympy import *

q1 = kf.get_vars("q^1", 4, numb=False)
q2 = kf.get_vars("q^2", 4, numb=False)
q1v = Matrix(q1[1:4])
q2v = Matrix(q2[1:4])

print("Скалярная часть:")
q1[0]*q2[0] - q1v.dot(q2v)

Скалярная часть:


q^1_0*q^2_0 - q^1_x*q^2_x - q^1_y*q^2_y - q^1_z*q^2_z

In [4]:
print("Векторная часть:")
q1v*q2[0] + q2v*q1[0] + q1v.cross(q2v)

Векторная часть:


Matrix([
[q^1_0*q^2_x + q^1_x*q^2_0 + q^1_y*q^2_z - q^1_z*q^2_y],
[q^1_0*q^2_y - q^1_x*q^2_z + q^1_y*q^2_0 + q^1_z*q^2_x],
[q^1_0*q^2_z + q^1_x*q^2_y - q^1_y*q^2_x + q^1_z*q^2_0]])

###### Старый код

In [40]:
import numpy as np

def vec2quat_old(v) -> list:
    if len(v) != 3:
        raise ValueError(f"Подаётся вектор длинны {len(v)}, требуется длинна 3!")
    if np.linalg.norm(v) > 1:
        return [0] + list(np.array(v)/np.linalg.norm(v))
    else:
        return [np.sqrt(1 - np.linalg.norm(v)**2), v[0], v[1], v[2]]

def q_dot_old(L1, L2):
    return np.array([L1[0] * L2[0] - L1[1] * L2[1] - L1[2] * L2[2] - L1[3] * L2[3],
                     L1[0] * L2[1] + L1[1] * L2[0] + L1[2] * L2[3] - L1[3] * L2[2],
                     L1[0] * L2[2] + L1[2] * L2[0] + L1[3] * L2[1] - L1[1] * L2[3],
                     L1[0] * L2[3] + L1[3] * L2[0] + L1[1] * L2[2] - L1[2] * L2[1]])

def quart2dcm_old(L) -> np.ndarray:
    w, x, y, z = L
    A = np.eye(3)
    A[0][0] = 1 - 2 * y ** 2 - 2 * z ** 2
    A[0][1] = 2 * x * y + 2 * z * w
    A[0][2] = 2 * x * z - 2 * y * w
    A[1][0] = 2 * x * y - 2 * z * w
    A[1][1] = 1 - 2 * x ** 2 - 2 * z ** 2
    A[1][2] = 2 * y * z + 2 * x * w
    A[2][0] = 2 * x * z + 2 * y * w
    A[2][1] = 2 * y * z - 2 * x * w
    A[2][2] = 1 - 2 * x ** 2 - 2 * y ** 2
    return A

def check_quat(q1, q2, q_dot, quart2dcm, vec2quat):
    print(f"λ₁ = {q1}")
    if isinstance(q1, list | np.ndarray):
        print(f"q3->q4: {vec2quat(q1[1:4])}")
        if isinstance(vec2quat(q1[1:4]), list | np.ndarray):
            print(f"модуль {vec2quat(q1[1:4])} = {np.linalg.norm(vec2quat(q1[1:4]))})")
    if isinstance(q1, quaternion.quaternion):
        q_new = vec2quat(q1.vec)
        print(f"q3->q4: {q_new}")
        print(f"модуль {q_new} = {q_new.abs()})")
    print(f"λ₂ = {q2}")
    print(f"λ₁∘λ₂ = {q_dot(q1, q2)}\n")

    A = quart2dcm(q1)
    print(f"A(λ₁) = \n{A} (type(A)={type(A)}) \nДетерминант {np.linalg.det(A)}, \nAA.T=\n{np.round(A @ A.T).astype(np.int64)}")

qs = [np.random.rand(4) for _ in range(2)]
for i in range(2):
    qs[i] /= np.linalg.norm(qs[i])

check_quat(qs[0], qs[1], q_dot_old, quart2dcm_old, vec2quat_old)

λ₁ = [0.63707903 0.43550425 0.60632635 0.19192373]
q3->q4: [0.6370790327569399, 0.43550424592350995, 0.6063263470304726, 0.19192373146841368]
модуль [0.6370790327569399, 0.43550424592350995, 0.6063263470304726, 0.19192373146841368] = 0.9999999999999999)
λ₂ = [0.79525458 0.07058782 0.50466811 0.32847771]
λ₁∘λ₂ = [0.10686248 0.49361367 0.67419132 0.53888034]

A(λ₁) = 
[[ 0.19106728  0.77265657 -0.60538841]
 [ 0.28357423  0.54700267  0.78763808]
 [ 0.93972281 -0.32216442 -0.11459117]] (type(A)=<class 'numpy.ndarray'>) 
Детерминант 0.9999999999999999, 
AA.T=
[[1 0 0]
 [0 1 0]
 [0 0 1]]


###### Новый код

In [41]:
for i in range(2):
    qs[i] = np.quaternion(*qs[i])

In [42]:
import numpy as np
import quaternion

def vec2quat(v):
    """Перевод вектора кватерниона в кватернион"""
    if isinstance(v, np.ndarray):
        return np.quaternion(np.sqrt(1 - np.linalg.norm(v)**2), *v)
    else:
        from sympy import Matrix, sqrt
        return Matrix([sqrt(1 - v.dot(v)), v[0], v[1], v[2]])

def q_dot(q1, q2):
    """Умножение кватернионов длины 4"""
    if isinstance(q1, quaternion.quaternion):
        return q1 * q2
    else:
        from sympy import Matrix
        q1v = Matrix(q1[1:4])
        q2v = Matrix(q2[1:4])
        scalar = q1[0]*q2[0] - q1v.dot(q2v)
        vector = q1v*q2[0] + q2v*q1[0] + q1v.cross(q2v)
        return Matrix([scalar, vector[0], vector[1], vector[2]])

def quart2dcm(q):
    """Матрицу поворота из кватерниона поворота"""
    if isinstance(q, quaternion.quaternion):
        return quaternion.as_rotation_matrix(q)
    else:
        from sympy import Matrix, sqrt
        w, x, y, z = q
        return Matrix([[1 - 2*y**2 - 2*z**2, 2*x*y + 2*z*w, 2*x*z - 2*y*w],
                       [2*x*y - 2*z*w, 1 - 2*x**2 - 2*z**2, 2*y*z + 2*x*w],
                       [2*x*z + 2*y*w, 2*y*z - 2*x*w, 1 - 2*x**2 - 2*y**2]]).T

check_quat(qs[0], qs[1], q_dot, quart2dcm, vec2quat)

λ₁ = quaternion(0.63707903275694, 0.43550424592351, 0.606326347030473, 0.191923731468414)
q3->q4: quaternion(0.63707903275694, 0.43550424592351, 0.606326347030473, 0.191923731468414)
модуль quaternion(0.63707903275694, 0.43550424592351, 0.606326347030473, 0.191923731468414) = 0.9999999999999999)
λ₂ = quaternion(0.795254582567478, 0.0705878228191371, 0.504668110019236, 0.328477711427569)
λ₁∘λ₂ = quaternion(0.106862484017262, 0.493613673207103, 0.674191317735446, 0.538880337573003)

A(λ₁) = 
[[ 0.19106728  0.28357423  0.93972281]
 [ 0.77265657  0.54700267 -0.32216442]
 [-0.60538841  0.78763808 -0.11459117]] (type(A)=<class 'numpy.ndarray'>) 
Детерминант 0.9999999999999999, 
AA.T=
[[1 0 0]
 [0 1 0]
 [0 0 1]]


**Копия старых результатов**

λ₁ = [0.63707903 0.43550425 0.60632635 0.19192373]
q3->q4: [0.6370790327569399, 0.43550424592350995, 0.6063263470304726, 0.19192373146841368]
модуль [0.6370790327569399, 0.43550424592350995, 0.6063263470304726, 0.19192373146841368] = 0.9999999999999999)
λ₂ = [0.79525458 0.07058782 0.50466811 0.32847771]
λ₁∘λ₂ = [0.10686248 0.49361367 0.67419132 0.53888034]

A(λ₁) = 
[[ 0.19106728  0.77265657 -0.60538841]
 [ 0.28357423  0.54700267  0.78763808]
 [ 0.93972281 -0.32216442 -0.11459117]] (type(A)=<class 'numpy.ndarray'>) 
Детерминант 0.9999999999999999, 
AA.T=
[[1 0 0]
 [0 1 0]
 [0 0 1]]

*ТАК а почему матрица транспонированная-то была?*

<u>Проверка матрицы из кватернионов</u><br>
Поворот на $90^\circ$ вокруг оси $Oz$: 
$$\lambda = [\cos\frac{\varphi}{2}, 0, 0, \sin\frac{\varphi}{2}]$$
$$x\to y, \hskip20px y \to -x$$

In [39]:
q1 = np.quaternion(1, 0, 0, 1).normalized()
A = quart2dcm(q1)
A = np.round(A).astype(np.int64)
print(f"A: \n{A}\n")

a = np.array([1, 0, 0])
print(f"x->y: {a} -> {A @ a}")

a = np.array([0, 1, 0])
print(f"y->-x: {a} -> {A @ a}")

A: 
[[ 0 -1  0]
 [ 1  0  0]
 [ 0  0  1]]

x->y: [1 0 0] -> [0 1 0]
y->-x: [0 1 0] -> [-1  0  0]


<u>Проверка всего того же с символьным вычислением</u><br>

In [47]:
from sympy import *

q1 = np.array([1, 2, 3, 4])
q2 = np.array([1, 2, 3, 4])

q1_n = np.quaternion(*q1).normalized()
q2_n = np.quaternion(*q2).normalized()
print(f"Численные кватернионы:\nq1 = {q1_n}\nq2 = {q2_n}\n")

q1_s = Matrix(q1 / np.linalg.norm(q1))
q2_s = Matrix(q2 / np.linalg.norm(q2))
print(f"Символьные кватернионы:\nq1 = {q1_s}\nq2 = {q2_s}")

Численные кватернионы:
q1 = quaternion(0.182574185835055, 0.365148371670111, 0.547722557505166, 0.730296743340221)
q2 = quaternion(0.182574185835055, 0.365148371670111, 0.547722557505166, 0.730296743340221)

Символьные кватернионы:
q1 = Matrix([[0.182574185835055], [0.365148371670111], [0.547722557505166], [0.730296743340221]])
q2 = Matrix([[0.182574185835055], [0.365148371670111], [0.547722557505166], [0.730296743340221]])
